### PR TITLE
feat: Snowflake support

### PR DESCRIPTION
**Pull request by mertmit**
## Change Summary

Added Snowflake support
- SnowflakeClient for NocoDB
- UI support for snowflake (SnowflakeUI)
- New ClientType: SNOWFLAKE

## Change type

- [x] feat: (new feature for the user, not a new feature for build script)

## Additional information / screenshots (optional)

image

### CODE REVIEW

File: packages/nocodb/src/lib/db/sql-data-mapper/lib/sql/BaseModelSqlv2.ts [hunk 1]
```diff
@@ -1808,7 +1819,19 @@ class BaseModelSqlv2 {
...
 
         if (ai) {
-          // response = await this.readByPk(id)
+          if (this.isSqlite) {
+            // sqlite doesnt return id after insert
+            id = (
+              await this.dbDriver(this.tnPath)
+                .select(ai.column_name)
+                .max(ai.column_name, { as: 'id' })
+            )[0].id;
```
Comment (pranavxc) [inline packages/nocodb/src/lib/db/sql-data-mapper/lib/sql/BaseModelSqlv2.ts:1828]: Are you sure that SQLite doesn't return the autoincremented primary key on insert?

Comment (mertmit) [reply packages/nocodb/src/lib/db/sql-data-mapper/lib/sql/BaseModelSqlv2.ts:1818]: Looks like knex introduced returning for sqlite3 in [v2.0.3](https://github.com/knex/knex/releases/tag/2.3.0)
I will cover this on another PR as this is a little bit out of scope

File: packages/nc-gui/components/dashboard/settings/data-sources/CreateBase.vue [hunk 2]
```diff
@@ -383,6 +392,43 @@ watch(
...
 
+      <template v-else-if="formState.dataSource.client === ClientType.SNOWFLAKE">
+        <!-- Account -->
+        <a-form-item label="Account" v-bind="validateInfos['dataSource.connection.account']">
+          <a-input v-model:value="formState.dataSource.connection.account" class="nc-extdb-account" />
+        </a-form-item>
+
+        <!-- Username -->
+        <a-form-item :label="$t('labels.username')" v-bind="validateInfos['dataSource.connection.user']">
+          <a-input v-model:value="formState.dataSource.connection.username" class="nc-extdb-host-user" />
```
Comment (wingkwong) [inline packages/nc-gui/components/dashboard/settings/data-sources/CreateBase.vue:403]: The validation rule name has to be matched with the property name. Currently, there is no validation on userame.

image

In L403, please change to

```
v-model:value="formState.dataSource.connection.user"
```

File: packages/nc-gui/components/dashboard/settings/data-sources/EditBase.vue [hunk 3]
```diff
@@ -376,6 +385,43 @@ onMounted(async () => {
...
 
+      <template v-else-if="formState.dataSource.client === ClientType.SNOWFLAKE">
+        <!-- Account -->
+        <a-form-item label="Account" v-bind="validateInfos['dataSource.connection.account']">
+          <a-input v-model:value="formState.dataSource.connection.account" class="nc-extdb-account" />
+        </a-form-item>
+
+        <!-- Username -->
+        <a-form-item :label="$t('labels.username')" v-bind="validateInfos['dataSource.connection.user']">
+          <a-input v-model:value="formState.dataSource.connection.username" class="nc-extdb-host-user" />
```
Comment (wingkwong) [inline packages/nc-gui/components/dashboard/settings/data-sources/EditBase.vue:396]: same as above

**Review by wingkwong — COMMENTED** [review]
1. Any dummy credential can be shared (in private)?

image

2. Is it allowed to add a new snowflake source from this menu?
image

3. `Use Connection URL` and `SSL & Advanced Parameters` are missing in create base page but shown in create external page.

**Pull request merged**